R/groupingsets.R

```diff
@@ -57,6 +57,16 @@ groupingsets.data.table = function(x, j, by, sets, .SDcols, id = FALSE, jj, ...)
     stopf("Argument 'sets' must be a list of character vectors.")
   if (!is.logical(id))
     stopf("Argument 'id' must be a logical scalar.")
+  if (!(is.null(total.label) ||
+        (is.character(total.label) && length(total.label) == 1L) ||
+        (is.list(total.label) && all(vapply_1b(total.label, is.character)) &&
```

Comment:
We can rule out list because AFAIU chsr vector will be sufficient. Pleasr correct me if I am wrong.

Reply:
Yes, as long as each element in the list is scalar then list is not necessary and character vector is sufficient.

Reply:
@jangorecki Thanks, I'll try that.

Reply:
I think I was wrong and labels must not be character type, right? Then we need a list type. Would be good to use such little bit  more complex example so it comes up straight away.

Reply:
@jangorecki No, currently the label values have to be character, and they're only applied to variables that are character or factor.

Reply:
Do you think total labels for other variable types should be allowed, and if so, which types?

For dates, I assume the user would be more likely to want a word like "Total" than to put in a date. For integers, they might want an integer code for the total, but if it was me I'd be more likely to want a word. I guess the function could convert the variable to character in the output, or leave it to the user to convert to character before using the function.

Reply:
It has to support whatever 'by' supports AFAIU. I should have some time this week so can look at it from my laptop. Recently I am on mobile only.

Reply:
Thanks @jangorecki. Currently it only supports character and factor grouping variables. For other types, it's left as NA.

In https://github.com/Rdatatable/data.table/issues/5351 you wrote
> Then we of course want total.label="Total" to be recycled to list(State = "Total", Group = "Total"). Of course only if all columns are character.

I interpreted this as meaning the label argument should only apply to variables of type character, but now I realise that maybe I interpreted incorrectly and you might have meant that `total.label = "Total"` should only be recycled to character variables, but non-character labels for non-character variables are still allowed in a list.

A couple of options:

1. Only apply the label argument to grouping variables of type character or factor, and require `total.label = c(Region = "National", Product = "Total")` instead of `total.label = list(Region = "National", Product = "Total")`.
2. Allow the label argument to apply to grouping variables that aren't character or factor, and allow total.label to be a list so that different types can be specified for different variables.

I prefer option 1, but I can try option 2 if you think that's better. As I mentioned above, if I had a date or integer grouping variable, I would probably still want a character total label (e.g. "Total") rather than a date or integer total label.

Reply:
Changing column type is not an option, column type needs to be controlled in j argument and not in labels (so no for option B). Option 2. We cannot apply it to all variables because we must match column types, we don't just print results but return a data.table so saying it is only for output won't be useful. Our output is data.table and column types cannot depend on labels arg. It is the labels arg has to match to column types in the output 'by'.

Reply:
By "output", I meant the data.table that is returned. Sorry for the confusion.

Thanks for clarifying that the column types have to stay the same. That makes it simpler.

Do you think the shorter form `label = "Total"` should be allowed, or should `label` always be a named list? If the shorter form is allowed, should it always have to be character, or would something like `label = 999L` be allowed and apply to all the integer grouping variables?

Reply:
Should be allows but for non-char column will result an NA, which anyway is likely to be expected.

Reply:
Thanks, I'll give that a try.